cpp/src/strings/json/json_path.cu

```diff
@@ -0,0 +1,957 @@
+  CUDA_HOST_DEVICE_CALLABLE parser(const char* _input, int64_t _input_len)
+    : input(_input), input_len(_input_len), pos(_input)
+  {
+    parse_whitespace();
+  }
+
+  CUDA_HOST_DEVICE_CALLABLE parser(parser const& p)
+    : input(p.input), input_len(p.input_len), pos(p.pos)
+  {
+  }
+
+  CUDA_HOST_DEVICE_CALLABLE bool eof(const char* p) { return p - input >= input_len; }
+  CUDA_HOST_DEVICE_CALLABLE bool eof() { return eof(pos); }
+
+  CUDA_HOST_DEVICE_CALLABLE bool parse_whitespace()
+  {
+    while (!eof()) {
+      if (is_whitespace(*pos)) {
+        pos++;
+      } else {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  CUDA_HOST_DEVICE_CALLABLE parse_result parse_string(string_view& str,
+                                                      bool can_be_empty,
+                                                      char quote)
+  {
+    str = string_view(nullptr, 0);
+
+    if (parse_whitespace() && *pos == quote) {
+      const char* start = ++pos;
+      while (!eof()) {
+        if (*pos == quote) {
+          str = string_view(start, pos - start);
+          pos++;
+          return parse_result::SUCCESS;
+        }
+        pos++;
+      }
+    }
+
+    return can_be_empty ? parse_result::EMPTY : parse_result::ERROR;
+  }
+
+  // a name means:
+  // - a string followed by a :
+  // - no string
+  CUDA_HOST_DEVICE_CALLABLE parse_result parse_name(string_view& name,
+                                                    bool can_be_empty,
+                                                    char quote)
+  {
+    if (parse_string(name, can_be_empty, quote) == parse_result::ERROR) {
+      return parse_result::ERROR;
+    }
+
+    // if we got a real string, the next char must be a :
+    if (name.size_bytes() > 0) {
+      if (!parse_whitespace()) { return parse_result::ERROR; }
+      if (*pos == ':') {
+        pos++;
+        return parse_result::SUCCESS;
+      }
+    }
+    return parse_result::EMPTY;
+  }
+
+  // numbers, true, false, null.
+  // this function is not particularly strong. badly formed values will get
+  // consumed without throwing any errors
+  CUDA_HOST_DEVICE_CALLABLE parse_result parse_non_string_value(string_view& val)
+  {
+    if (!parse_whitespace()) { return parse_result::ERROR; }
+
+    // parse to the end of the value
+    char const* start = pos;
+    char const* end   = start;
+    while (!eof(end)) {
+      char const c = *end;
+      if (c == ',' || c == '}' || c == ']' || is_whitespace(c)) { break; }
+
+      // illegal chars
+      if (c == '[' || c == '{' || c == ':' || c == '\"') { return parse_result::ERROR; }
+      end++;
+    }
+    pos = end;
+
+    val = string_view(start, end - start);
+
+    return parse_result::SUCCESS;
+  }
+
+ protected:
+  char const* input;
+  int64_t input_len;
+  char const* pos;
+
+ private:
+  CUDA_HOST_DEVICE_CALLABLE bool is_whitespace(char c)
+  {
+    return c == ' ' || c == '\r' || c == '\n' || c == '\t' ? true : false;
```

Comment:
I think it would be safe to shorten this to `c <= ' '`. 
Characters below space in the ascii table are generally unusable or control characters like these.